clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -591,6 +591,11 @@ void pollSelectionKeys(Set<SelectionKey> selectionKeys,
                 long nowNanos = channelStartTimeNanos != 0 ? channelStartTimeNanos : currentTimeNanos;
                 try {
                     attemptWrite(key, channel, nowNanos);
+
+                    // Following is to fix KAFKA-13559. This will prevent poll() from blocking for 300 ms when the
+                    // socket has no data but the buffer has data. Only happens when using SSL.
+                    if (channel.hasBytesBuffered())
+                        madeReadProgressLastPoll = true;
```

Comment:
This is probably not the right place for this logic - we only want to set the flag if we completed a send and updated the mute state for a channel.

@ijuma's comment was probably more around the naming being unintuitive - we were previously setting `madeReadProgressLastPoll` when completing a Send.

We probably need a clarifying comment around why completing a `Send` is making read progress - there is an eligible channel to read from that previously wasn't available. 

Alternatively, we can update the variable name to something that captures the scenario we're trying to cover (or add a new variable). Maybe something like `addedNewBufferedKeysLastPoll`. It's not very succinct but I've always been terrible at variable naming.

Reply:
@splett2 

How about saying this in the comment:

```
After reading from the channel, "madeReadProgressLastPoll" was set to "true". However, because channel is muted while processing the previous request, "madeReadProgressLastPoll" was set to "false" by "clear()". Now that a response has been sent and channel is unmuted, we need to set "madeReadProgressLastPoll" back to "true" if there is additional data in the buffer, to ensure we process it immediately.
```

Reply:
@splett2 

The `Selector` method getting called after channel is unmuted is actually `clearCompletedSends()` ([here](https://github.com/apache/kafka/blob/a6c036256504b2549b009c10c777927a40a4bc34/clients/src/main/java/org/apache/kafka/common/network/Selector.java#L825-L827)).

Few alternatives I can think of:

1. Move `madeReadProgressLastPoll=true` to this method (`Selector.clearCompletedSends()`).
2. Create a new `Selector` method named `readyForNextRead()` (or something) and call it from `SocketServer.processCompletedSends()`.
3. Not setting `madeReadProgressLastPoll=false` if `channel.hasBytesBuffered()` is true in `clear()` ([here](https://github.com/apache/kafka/blob/a6c036256504b2549b009c10c777927a40a4bc34/clients/src/main/java/org/apache/kafka/common/network/Selector.java#L862)).

What do you think?

Reply:
@badaiaqrandista Couldn't we just set madeReadProgressLastPoll=true in `unmute(KafkaChannel channel)` when we add the channel to keysWithBufferedRead? That is the case we are trying to handle here right? It would be more obvious why, without having to rely on comments.

```
if (channel.hasBytesBuffered()) {
     keysWithBufferedRead.add(channel.selectionKey());
     madeReadProgressLastPoll = true;
}
```

Reply:
Thank you. You are right. And the test passed.